### PR TITLE
Story 2 - Prompt and first test runs

### PR DESCRIPTION
**Pull request by jefedigital**
[Aug 9 - this PR is in Draft mode .. going to breakdown the edit suggestions into modules and push associated branches so we can collab]

Implemented a notebook to run our first prompt drafts to extract concepts, interventions, and edges from the first test paper, "Deception Abilities Emerged in Large Language Models" (Hagendorff 2023), using the o3 model. 

The full prompt produced 11 concepts, 5 interventions, and 12 edges. Running targeted prompts for concepts and interventions separately yielded an additional +6 concepts and +1 intervention, indicating improved coverage with focused extraction.

- Notebook: **Story2/entity_extraction.ipynb**
- Test Run Outputs: **Story2/outputs**
- Prompt definitions: **Story2/prompts.py**
- Schema definitions: **Story2/schemas.py**

Next steps:

- Appending dedicated Concepts and Intervention node JSONs and prompting model to focus on Edge detection only. 
- Experiment with simple "self-reflection" loop forcing model to iterate and identify / refine more causal relationships (referenced in Tong).
- Experiment with few-shot examples (none given)

Other observations: 

- Didn't appear to need to call structured output tools, passing JSON schema and giving clear instructions in the prompt seemed sufficient.

### CODE REVIEW

File: Story2/prompts.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  "nodes": [
+    {
+      "node_name": "<UPPER-SNAKE-CASE TERM>",
+      "node_description": "<VERBOSE TERM DESCRIPTION>",
+      "isIntervention": 0 | 1,  // 0 = Concept, 1 = Intervention
+
+      // include BOTH of these ONLY when isIntervention == 1
+      "pipeline_stage": <integer 0-5>   // 0 pre-training … 5 other
+      "maturity_level": <integer 1-5>   // 1 theory-only … 5 verified on 100B+ LLM
+    },
```
Comment (MartinLeitgab) [inline Story2/prompts.py:27]: We should include more detailed rubrics for pipeline and maturity level, we can develop those with AI help.

File: Story2/prompts.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+NO OTHER FIELDS ARE ALLOWED.  
+
+
+----------  EXTRACTION STEPS (internal)  ----------
+Work through the paper in the input using the following internal reasoning steps.
+
+1. **Concept Node sweep** – Scan abstract, introduction, related-work, methods, discussion, and conclusion for key *problems, ideas, or theoretical constructs* and create a NODE object for each. 
+  • Normalise each node_name to UPPER-SNAKE-CASE.
```
Comment (MartinLeitgab) [inline Story2/prompts.py:50]: We can instruct the LLM to focus on e.g. 'key thought concepts that describe a logical chain towards the proposed interventions in this paper'. That way the LLM may not get caught up in thinking of more detailed categories like problems etc that may then make it harder to merge across papers in the next workflow step.

File: Story2/prompts.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+----------  EXTRACTION STEPS (internal)  ----------
+Work through the paper in the input using the following internal reasoning steps.
+
+1. **Concept Node sweep** – Scan abstract, introduction, related-work, methods, discussion, and conclusion for key *problems, ideas, or theoretical constructs* and create a NODE object for each. 
+  • Normalise each node_name to UPPER-SNAKE-CASE.   
+  • Merge obvious synonyms in node_name; but keep distinct concepts if the paper treats them separately. 
+  • Write a concise but informative node_description (1-2 sentences) for each concept, based on the paper’s own wording.
+
+2. **Intervention Node sweep** – Scan methods, experiments, results, evaluations for any *method, treatment, algorithm, dataset, training trick, measurement, or policy* proposed or empirically tested and create a NODE object for each.
+  • Normalise each node_name to UPPER-SNAKE-CASE.
```
[the remark this answers is not in the document]
Comment (MartinLeitgab) [reply Story2/prompts.py:55]: We should also add that interventions should form the endpoints of the logical chains to be created.

**Review by MartinLeitgab — COMMENTED** [review]
This is great, thank you @jefedigital for creating and sharing.

- [ ] Token efficiency: Good insights already gained on focused extraction improvement opportunity- for input token economics/cost limitation it would be great to only pass the input paper once over the API and then let the model iterate over focused extraction, reflection etc. (this may already be implemented this way). Also to save some tokens we maybe can look into cutting off the paper at the beginning of the bibliography/references section such that we do not submit the full paper to o3, this may save us some cost. The bibliography and appendices often do not critically contribute to intervention chains and we can leave those out.
- [ ] Maybe can save 'Alias' datamembers: If the 'aliases' data members are intended as bootstrap for later global Intervention Graph merging, we maybe can save/not allocate since merging LLM expected to be capable enough (reasoning) to merge sufficiently similar nodes in similar context.
- [ ] Maturity level and pipeline stage rubric: We should invest in creating a more robust rubric for maturity_level and stage_in_pipeline with LLM help offline, maybe @AgastyaMalik is already working on this.
- [ ] Specificity in concept/edge nodes/titles: One element we should focus on is specificity in concepts and interventions, and emphasize that to the LLM in the prompt. E.g. a 'Scalable Automated Alignment Framework' is a very large, multi-layered, and complex endeavor compared to a specific intervention that targets to alleviate a specific issue. If the paper explicitly points to this intervention that is ok, but we should emphasize to the LLM that specificity is a key attribute for both concepts and especially interventions.
- [ ] Rationale records for explainability: For explainability reasons, we want to log rationale by the LLM on each workflow step it takes, but we do not necessarily need to store this in the main intervention graph/could store this offline as long as it can be retrieved from the associated attributes. As design options we e.g. can prompt the LLM to create a rationale record for all information generated (concepts/interventions, edges, and their datamembers) and store that record in a separate database that is linked to e.g. the DOI of the paper. Then any end user who wants to learn more why a certain edge, or concept data member, or intervention shows up in the intervention graph as it does, can retrieve that record by the DOI link and find all rationale the LLM has applied.  This also can apply to later processing steps like global graph node/edge merging, and later intervention evaluation and graph traversal, e.g. the rationale record could be loaded into the context of the merging LLM to consider the context of nodes to be merged (there may be a resource/token trade-off). We can have more discussion around this.

Some other observations:
- Good find that structured output tools may not be needed.
- I added additional in-line comments in prompts.py for specific sections in the prompt_full string, which also apply by intent to the prompt_concepts and prompt_edges strings. 
- I had to educate myself a bit what may work to extract chains per my own requirements- I got inputs from Claude drafted into a prompt sketch, see link below, that I fed into Gemini together with our dev paper. This produced a maybe reasonable graph as linked below as well (not fully trusting Gemini's visualization skills). There is a few inputs that we can probably adopt for robustness of nodes/edges and chains, including:

1. We may need to still need to start with problem statements to have the LLM embrace the task of constructing a logical chain towards interventions (including 'toy' JSON outputs to get the LLM to commit to this framing, where we may just scrap those outputs in merging the local graphs)
2. How to capture both the broader ideas in a given paper and the specific context the paper is picking out (via 'refined by' edges in neighboring concepts)
3. Still need to include some descriptions of nodes and edges to take along more context (with more detailed rationale in the processing log connected with the intervention graph elements through the paper DOI)
4. Manage papers which are just reviews of a subfield with potentially many interventions but short chains, as well as papers outside of AI safety which however we should still harvest to cross-pollinate possible interventions (but flagging such inferences by the LLM as not actually proposed in the paper)
5. Rubrics for maturity, lifecycle stage (not included in the prompt sketch below), and edge confidence
6. Few shot examples for edge and concept/intervention texts

Please see the prompt sketch I created, and then outputs from Gemini 2.5 Pro for the same paper that Jeff worked on (2307.16513v2) in the folder on our shared Google drive (including rationale record and JSON output), https://drive.google.com/drive/folders/18eaiDmE6X7scmayuNorvLQ70vAp69iMJ?usp=drive_link .

File: Story2/prompts.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+  • Estimate confidence from the paper’s own wording:  
+    – speculative / future work ⇒ 1–2  
+    – correlation / qualitative evidence ⇒ 3  
+    – controlled experiment / quantified metrics ⇒ 4–5
+
+4. **Validation sweep** – Ensure:  
+  • Every source/target in edges exists in nodes.  
+  • No duplicate nodes (same node_name) in the results. 
+  • JSON is syntactically valid and matches the schema exactly.
+
```
Comment (MartinLeitgab) [inline Story2/prompts.py:73]: In validation we should also check for unconnected nodes- with prompt updates towards focusing on logical chains leading to interventions, we should not have satellite/unconnected nodes remaining.
Intervention_Graph_Full

File: Story2/prompts.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+IMPORTANT RULES  
+  • Return *only* the JSON object shown above – no commentary or markdown.
+  • Ensure valid JSON syntax (e.g. double quotes, no trailing commas, no curly quotes).
+  • Do **not** include bibliographic metadata (DOI, authors, institutions, dates).  
+  • Leave arrays empty when they have no elements.  
+  • No extra fields or top-level wrapper arrays.
+
+END OF INSTRUCTIONS
+"""
```
Comment (MartinLeitgab) [inline Story2/prompts.py:249]: An input for an updated full_prompt draft is below:

SYSTEM:
You are a meticulous scientific information-extraction assistant.

Read the attached research paper.


----------  ANALYSIS INSTRUCTION STEPS  ----------
Work through the paper in the input using the following reasoning steps. Please output rationale text for **every** concept/intervention and edge identified, as well as for their data member values you choose.

1. **Concept and Intervention Node sweep** – Think carefully and scan all major sections of the paper (e.g. abstract, introduction, related-work, methods, discussion, and conclusion) to extract key thought concepts that allow to form logical chains towards the proposed interventions in this paper. Interventions should be the end point of their respective logical chains. Interventions are defined as a specific change to the current model development paradigm at frontier model development companies proposed by the authors to resolve the identified problem, issue, or improvement opportunity by this paper. Create a NODE object for each key thought concept and intervention. 
  • Normalise each node_name to UPPER-SNAKE-CASE.   
  • Merge obvious synonyms in node_name; but keep distinct concepts if the paper treats them separately. 
  • Write a concise but informative node_description (1-2 sentences) for each concept, based on the paper’s own wording.
  • Decide whether isIntervention = 1.  
  • Infer pipeline_stage and maturity_level from context (e.g. “during fine-tuning we…”, “tested on 7B model…”, “deployed in product…”). 

2. **Edge sweep** – Think carefully and from the content of the paper, create an EDGE object for each logical connection between concepts, and between the last concept preceding the intervention and the intervention itself. 
  • Normalise each edge_name to UPPER-SNAKE-CASE.
  • Write a concise but informative edge_description (1-2 sentences) for each edge, based on the paper’s own wording.
  • Direction always points from *cause / prerequisite / antecedent* → *effect / next step / outcome*.  
  • Estimate confidence from the paper’s own wording:  
    – speculative / future work ⇒ 1–2  
    – correlation / qualitative evidence ⇒ 3  
    – controlled experiment / quantified metrics ⇒ 4–5

3. **Validation sweep** – Ensure:  
  • Every source/target in edges exists in nodes.  
  • No duplicate nodes (same node_name) in the results. 
  • No satellite nodes or nodes without edges to other nodes in the results.
  • JSON is syntactically valid and matches the schema exactly.

----------  END OF ANALYSIS INSTRUCTION STEPS  ----------

---------- RATIONALE PRINTOUT STEP --------------

Please print all rationale output generated for each of the concepts/interventions and edges, including all of their data members.

---------- END OF RATIONALE PRINTOUT STEP --------------


---------- JSON FORMAT DATA PRINTOUT STEP --------------

At the end of your output, please return a section of text that includes every
  • Concept node
  • Intervention node
  • Edge that links them
in a strict JSON schema, using the analysis steps you have performed. 

STRICT OUTPUT:  
Return **only** valid JSON (no markdown) with the following top-level structure:

{
  "nodes": [
    {
      "node_name": "<UPPER-SNAKE-CASE TERM>",
      "node_description": "<VERBOSE TERM DESCRIPTION>",
      "isIntervention": 0 | 1,  // 0 = Concept, 1 = Intervention

      // include BOTH of these ONLY when isIntervention == 1
      "pipeline_stage": <integer 0-5>   // 0 pre-training … 5 other
      "maturity_level": <integer 1-5>   // 1 theory-only … 5 verified on 100B+ LLM
    },
    ...
  ],

  "edges": [
    {
      "edge_name": "<UPPER-SNAKE-CASE RELATION NAME>",
      "edge_description": "<VERBOSE RELATION DESCRIPTION>",
      "source_node": ["<node_name>", ...],   // must match nodes[].node_name exactly
      "target_node": ["<node_name>", ...],
      "confidence"  : <integer 1-5>.  // 1 speculative → 5 strong evidence
    },
    ...
  ]
}

NO OTHER FIELDS ARE ALLOWED IN THIS SECTION.  

IMPORTANT RULES  
  • In the final section return *only* the JSON object shown above – no commentary or markdown.
  • Ensure valid JSON syntax (e.g. double quotes, no trailing commas, no curly quotes).
  • Do **not** include bibliographic metadata (DOI, authors, institutions, dates).  
  • Leave arrays empty when they have no elements.  
  • No extra fields or top-level wrapper arrays.

---------- END OF JSON FORMAT DATA PRINTOUT STEP --------------

File: Story2/prompts.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+You are a meticulous scientific information-extraction assistant.
+
+Read the provided research paper and return every
+  • Concept node
+  • Intervention node
+  • Edge that links them
+in a strict JSON schema.
+
+
+STRICT OUTPUT:
```
[the remark this answers is not in the document]
Comment (jonpsy) [reply Story2/prompts.py:14]: Agreed

File: Story2/schemas.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
+SCHEMA_FULL = {
```
Comment (jonpsy) [inline Story2/schemas.py:1]: This is great. I think defining schema in pydantic might be more suitable.
